sph/include/sph/cooling.hpp

```diff
@@ -0,0 +1,141 @@
+    gr_float HDI_density                   = (gr_float)d.HDI_fraction[i] * (gr_float)d.rho[i];
+    gr_float metal_density                 = (gr_float)d.metal_fraction[i] * (gr_float)d.rho[i];
+    grackle_fields.HI_density              = &HI_density;
+    grackle_fields.HII_density             = &HII_density;
+    grackle_fields.HeI_density             = &HeI_density;
+    grackle_fields.HeII_density            = &HeII_density;
+    grackle_fields.HeIII_density           = &HeIII_density;
+    grackle_fields.e_density               = &e_density;
+    grackle_fields.HM_density              = &HM_density;
+    grackle_fields.H2I_density             = &H2I_density;
+    grackle_fields.H2II_density            = &H2II_density;
+    grackle_fields.DI_density              = &DI_density;
+    grackle_fields.DII_density             = &DII_density;
+    grackle_fields.HDI_density             = &HDI_density;
+    grackle_fields.metal_density           = &metal_density;
+    gr_float vhr                           = 0.;
+    gr_float shr                           = 0.;
+    gr_float HI_ir                         = 0.;
+    gr_float HeI_ir                        = 0.;
+    gr_float HeII_ir                       = 0.;
+    gr_float H2_dr                         = 0.;
+    gr_float hr                            = 0.;
+    grackle_fields.volumetric_heating_rate = &vhr;
+    grackle_fields.specific_heating_rate   = &shr;
+    grackle_fields.RT_HI_ionization_rate   = &HI_ir;
+    grackle_fields.RT_HeI_ionization_rate  = &HeI_ir;
+    grackle_fields.RT_HeII_ionization_rate = &HeII_ir;
+    grackle_fields.RT_H2_dissociation_rate = &H2_dr;
+    grackle_fields.RT_heating_rate         = &hr;
+    gr_float isrf                          = 1.7;
+    grackle_fields.isrf_habing             = &isrf;
+
+    // std::cout << grackle_units_copy.density_units << std::endl;
+    // std::cout << d.dt[i] / grackle_units.time_units << std::endl;
+    // std::cout << HI_density / gr_rho << std::endl;
+    // std::cout << HeI_density / gr_rho << std::endl;
+    // std::cout << gr_u << std::endl;
+
+    solve_chemistry(&grackle_units_copy, &grackle_fields, d.dt[i] / grackle_units.time_units);
+    // std::cout << HI_density/gr_rho << std::endl;
+    d.rho[i]            = gr_rho;
+    d.HI_fraction[i]    = HI_density / gr_rho;
+    d.HII_fraction[i]   = HII_density / gr_rho;
+    d.HeI_fraction[i]   = HeI_density / gr_rho;
+    d.HeII_fraction[i]  = HeII_density / gr_rho;
+    d.HeIII_fraction[i] = HeIII_density / gr_rho;
+    d.e_fraction[i]     = e_density / gr_rho;
+    d.HM_fraction[i]    = HM_density / gr_rho;
+    d.H2I_fraction[i]   = H2I_density / gr_rho;
+    d.H2II_fraction[i]  = H2II_density / gr_rho;
+    d.DI_fraction[i]    = DI_density / gr_rho;
+    d.DII_fraction[i]   = DII_density / gr_rho;
+    d.HDI_fraction[i]   = HDI_density / gr_rho;
+    d.metal_fraction[i] = metal_density / gr_rho;
+    d.u[i]              = gr_u;
+}
```

Comment:
Does ChanGa also convert between fractions and densities every particle and iteration?

Reply:
Yes.